tools/ExtensionDownloader.ps1

```diff
@@ -0,0 +1,216 @@
+function print-objectJson
+{
+    param($objectJson);
+
+    write-host "downloadLocation:" $objectJson.downloadLocation;
+    write-host "FileHash:" $objectJson.FileHash;
+    write-host "Hash-Algorithm:" $objectJson.HashAlgorithm;
+    write-host "CodeSignPolicyFile:" $objectJson.codeSignPolicyFile;
+}
+
+# ===================================================================
+#
+#   Main entry point.
+#
+# ===================================================================
+function Main
+{
+    print-EnvironmentVariables;
+
+    $extensionJson = get-ChangedExtensions;
+
+   if($null -ne $extensionJson)
+    {
+        $extensionRootDirectory = (get-BuildRootDirectory); 
+
+        write-host "Repository root:" $extensionRootDirectory;
+
+        foreach($json in $extensionJson)
+        {
+            $jsonFile = $extensionRootDirectory + "\" + $json;
+
+            Write-Host "Processing extension json:" $jsonFile;
+
+            $objectName = [System.IO.Path]::GetFileNameWithoutExtension($jsonFile);
+
+            $objectInfo = Get-Content $jsonFile | ConvertFrom-Json;
+            
+            print-objectJson -objectJson $objectInfo;
+
+            $pFile  = ${env:Build_SourcesDirectory} + "\objects\ConsoleExtension\" + $objectInfo.codeSignPolicyFile;
+            Write-Host "##vso[task.setvariable variable=codeSignPolicyFile;]$pFile"
+
+            $itemDir = $extensionRootDirectory + "\objects\consoleextension\" + $objectName;
+            $cabFile = $itemDir + "\" + $objectName + ".cab"
+
+            $r = mkdir $itemDir;
+    
+            if((Test-Path $cabFile) -eq $False )
+            {
+                Write-Host "Downloading cab:" $objectInfo.downloadLocation "to:" $itemDir;
+                Invoke-WebRequest -Uri $objectInfo.downloadLocation -OutFile $cabFile;
+            }
+
+            verifyFileHash -expectedHash $objectInfo.FileHash -fileToCheck $cabFile -algorithm $objectInfo.HashAlgorithm -ErrorAction Stop
+    
+            write-host "Recursively searching for cab files.."
+            searchAndExpand -directory $itemDir
+
+            print-Summary;
+        }
+    }
+    else {
+        write-host "No extensions submitted to this branch.";
+    }
+}
+
+# ===================================================================
+#
+#   Gets the hash of the specified file and compares it to the expected hash
+#
+# ===================================================================
+function verifyFileHash
+{
+    param($expectedHash, $fileToCheck, $algorithm);
+
+    Write-Host "Verifying hash of file:" $fileToCheck;
+
+    $actualHash =  Get-FileHash -Path $fileToCheck -Algorithm $algorithm;
+    
+    write-host "Algorithm:" $actualHash.Algorithm;
+    write-host "Path:" $actualHash.Path;
+    write-host "ExpectedHash: ["$expectedHash"] ActualHash ["$actualHash.Hash"].";
+
+    if ($actualHash.Hash -ne $expectedHash)
+    {
+        Write-Error $message;
```

Comment:
Is Write-Error enough to make this fail here?  I think you need to throw don't you?

Reply:
It actually is.... I had the same question so pushed this commit https://github.com/microsoft/configmgr-hub-selfhost/pull/623/commits/09f6b2a60524bac8414e1668fb8d14d738574d26 to check it and the pipeline failed. I'd give you a link to that pipeline if I could figure out how to match that run to that commit.  I figure you trust me though so I stopped looking for it.

Reply:
Oh, and I figured out how to find the check that was performed that was linked to the commit where I was testing that.  Here it is

https://github.com/microsoft/configmgr-hub-selfhost/pull/623/checks?sha=09f6b2a60524bac8414e1668fb8d14d738574d26